ports/esp32/esp32_sdcard.c

```diff
@@ -0,0 +1,386 @@
+
+        DEBUG_printf("  Calling init_slot()");
+        check_esp_err(sdmmc_host_init_slot(self->host.slot, &slot_config));
+    }
+
+    DEBUG_printf("  Returning new card object: %p", self);
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC mp_obj_t sd_deinit(mp_obj_t self_in) {
+    sdcard_card_obj_t *self = self_in;
+
+    DEBUG_printf("De-init host\n");
+
+    if (self->flags & SDCARD_CARD_FLAGS_HOST_INIT_DONE) {
+        self->host.deinit();
+        self->flags &= ~SDCARD_CARD_FLAGS_HOST_INIT_DONE;
+    }
+
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(sd_deinit_obj, sd_deinit);
+
+STATIC mp_obj_t sd_info(mp_obj_t self_in) {
+    sdcard_card_obj_t *self = self_in;
+    // We could potential return a great deal more SD card data but it
+    // is not clear that it is worth the extra code space to do
+    // so. For the most part people only care about the card size and
+    // block size.
+
+    check_esp_err(sdcard_ensure_card_init((sdcard_card_obj_t *) self, false));
+
+    uint32_t log_block_nbr = self->card.csd.capacity;
+    uint32_t log_block_size = _SECTOR_SIZE(self);
+
+    mp_obj_t tuple[3] = {
+        mp_obj_new_int_from_ull((uint64_t)log_block_nbr * (uint64_t)log_block_size),
+        mp_obj_new_int_from_uint(log_block_size),
+    };
+    return mp_obj_new_tuple(3, tuple);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(sd_info_obj, sd_info);
+
+STATIC mp_obj_t esp32_sdcard_readblocks(mp_obj_t self_in, mp_obj_t block_num, mp_obj_t buf) {
+    sdcard_card_obj_t *self = self_in;
+    mp_buffer_info_t bufinfo;
+    esp_err_t err;
+
+    err = sdcard_ensure_card_init((sdcard_card_obj_t *) self, false);
+    if (err != ESP_OK)
+        return false;
+
+    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_WRITE);
+    err = sdmmc_read_sectors(&(self->card), bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / _SECTOR_SIZE(self) );
+
+    return mp_obj_new_bool(err == ESP_OK);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(esp32_sdcard_readblocks_obj, esp32_sdcard_readblocks);
+
+STATIC mp_obj_t esp32_sdcard_writeblocks(mp_obj_t self_in, mp_obj_t block_num, mp_obj_t buf) {
+    sdcard_card_obj_t *self = self_in;
+    mp_buffer_info_t bufinfo;
+    esp_err_t err;
+
+    err = sdcard_ensure_card_init((sdcard_card_obj_t *) self, false);
+    if (err != ESP_OK)
+        return false;
+
+    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_READ);
+    err = sdmmc_write_sectors(&(self->card), bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / _SECTOR_SIZE(self) );
+
+    return mp_obj_new_bool(err == ESP_OK);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(esp32_sdcard_writeblocks_obj, esp32_sdcard_writeblocks);
+
+STATIC mp_obj_t esp32_sdcard_ioctl(mp_obj_t self_in, mp_obj_t cmd_in, mp_obj_t arg_in) {
+    sdcard_card_obj_t *self = self_in;
+    esp_err_t err = ESP_OK;
+    mp_int_t cmd = mp_obj_get_int(cmd_in);
+
+    switch (cmd) {
+        case BP_IOCTL_INIT:
+            err = sdcard_ensure_card_init(self, true);
```

Comment:
This is problematic: the FAT FS driver will call `BP_IOCTL_INIT` very often (at least every time a file is opened) so forcing re-initialisation of the SD card each time a file is opened will make it slow (I'm guessing, but I think init should only be done once at the start).  Probably worth putting in a printf here to test this out, mount a card, open a file, and see how often INIT is called.

Reply:
Aha. I wasn't sure of the semantics of that call but I was assuming it was only going to call this when the card was mounted and it would be important to make sure that we had the right card size information in case the user had swapped the card since it was last called. I will add a debug print statement and see how often it is called.

Reply:
I'm not sure how to handle swapping of cards, but I really don't think this function should be called each time there is a BP_IOCTL_INIT request.

Reply:
I've changed the code to ensure that the card is inited but to not repeatedly init the card.